### DISCUSSION TITLE
Counting yarn bags/cartons on pallets with YOLOv8/v11-seg with small dataset, heavy occlusion, dual fixed cameras

### DISCUSSION THREAD

**Discussion by Muavia-Shakeel** [ultralytics/ultralytics#25605] [Solutions]
We're building a warehouse pipeline (DeepStream/TensorRT) to count yarn **bags** (soft poly sacks) and **cartons** (rigid boxes) per pallet as they move through an aisle - 1 to 10 items per pallet, no fixed formula. Two fixed Dahua cameras sit at opposite ends of the same aisle (can't reposition); a pallet is seen first at one end, then the other, effectively showing **opposite faces** of the same stack.

Challenges: two stacking patterns (block vs. interlock - very different occlusion levels), occlusion from bag-on-bag, workers, and forklift masts, and a small annotated dataset so far (~150-300 images planned, COCO-seg via Roboflow).

Current plan: YOLOv8s-seg + BoT-SORT/StrongSORT tracker, amodal-style segmentation for occlusion, cross-camera count fusion to avoid double-counting the same pallet.

**Looking for advice on:**
1. Training recipe for small (~150-300 img), occlusion-heavy, 2-class seg datasets.
2. Best way to label/train for *amodal* (full-extent-even-if-occluded) instances in YOLO-seg.
3. Fusing counts from two sequential, opposite-facing (non-overlapping) camera views of the same object without double-counting.
4. Whether YOLOv8-seg is the right call here vs. something else, given edge/TensorRT deployment constraints.

Happy to share sample frames/annotations if useful.
@glenn-jocher @RizwanMunawar

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Muavia-Shakeel, thank you for sharing this detailed warehouse vision use case 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples, along with answers to many common questions.

For your occlusion-heavy custom training ❓ question, please provide as much information as possible, including representative dataset images, annotation examples, class definitions, camera viewpoints, model configuration, training logs, validation results, and your target TensorRT/DeepStream hardware. Please also verify that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips). If this is a 🐛 bug report, please include a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response 🤖, and an Ultralytics engineer will also assist you soon.

**Comment by ZC502** [comment]
@Muavia-Shakeel
This looks like a very interesting (and very real) warehouse vision problem. Edge deployment is exactly where debugging becomes difficult.
Regarding your point #4 (TensorRT / DeepStream deployment), there is one additional failure mode worth checking before spending too much effort on dataset refinement.
When a model is trained and evaluated in FP32, but later deployed through an optimized backend (TensorRT FP16/INT8, etc.), the exported model may not always produce identical outputs. This is especially important for borderline cases — for example:
*objects detected with confidence just above the threshold in FP32 may disappear after deployment;
*small changes in box/mask geometry may affect tracker IoU matching;
*downstream components like BoT-SORT may amplify small detection differences into ID switches.

So before assuming the counting error comes from annotation quality or the amodal labeling strategy, it may be useful to run a simple cross-backend parity check:
FP32 reference output vs TensorRT/DeepStream output on the same raw frames.
The goal is to answer:
"Is the model actually failing, or is the deployment pipeline changing the model output?"
I am currently exploring this exact problem space, focusing on automatically identifying deployment drift patterns such as confidence threshold crossing and geometry shifts. I actually just brought up this missing piece in the Ultralytics edge feedback loop discussion here: https://github.com/orgs/ultralytics/discussions/25250#discussioncomment-17851072

For your case, I would be especially interested in checking:
*segmentation confidence changes after export;
*mask/box IoU drift between backends;
*whether tracker instability starts only after deployment.

Good luck with the pipeline — this is exactly the kind of real-world edge case where deployment debugging tools are still missing in the ecosystem, but it is exactly what we are actively trying to solve right now.

**Comment by Muavia-Shakeel** [comment]
@ZC502 
Thanks, this is really useful  hadn't considered that the deployment backend itself could be the source of instability rather than the model/data. Makes sense that BoT-SORT would amplify small geometry/confidence shifts into ID switches, especially since we're already dealing with heavy occlusion where detections are naturally borderline-confidence to begin with.
We're still at the labeling stage right now (~150-300 seed images planned, only a handful annotated so far), so no trained model or export to test yet. But I'll bookmark this exact check for once we have a first trained YOLOv8-seg model and export to TensorRT will run the FP32 vs TensorRT parity comparison you described (confidence deltas, mask/box IoU drift, tracker stability with each backend's outputs fed separately) before assuming any count errors are a dataset/annotation problem.
Will check out the discussion thread you linked too. If we hit this once we're at deployment, I'll come back and report what we found could be useful data point for what you're exploring.
Thanks again for flagging this before we sink more time into dataset refinement.

**Reply by ZC502** [reply]
Thanks for the detailed follow-up!One small note from my experience: the parity check itself is usually straightforward, but the tricky part is isolating where the divergence starts (preprocessing, inference, decoding/post-processing, or tracking input).
For example, a confidence drop after TensorRT export does not always mean INT8 calibration is the cause — sometimes the first mismatch appears earlier in preprocessing or output decoding. Anyway, you are still at the right stage now: get the first FP32 baseline working, then keep a small fixed validation set for backend comparison. That will make the later debugging much easier.
Looking forward to hearing what you find when you reach deployment!

**Comment by rishig47-dev** [comment]
For a setup like this, I would simplify the pipeline as much as possible because you have only ~150–300 images and two fixed cameras.

For training, I would start from a pretrained `yolov8s-seg.pt` or `yolo11s-seg.pt` rather than training from scratch. With a small dataset, focus heavily on representative annotations and augmentation rather than simply increasing epochs. Your train/validation split should also be done by pallet/sequence rather than randomly splitting consecutive frames, otherwise validation leakage can make the results look better than they really are. Ultralytics recommends pretrained segmentation weights for custom training.

For the occlusion problem, a modal masks can work if you consistently annotate the estimated full extent of each bag/carton, including the hidden region. However, the hidden boundary is inherently ambiguous, so I would first establish whether visible-instance masks are sufficient. If the downstream requirement is counting, a robust detector + tracker may be more valuable than trying to make every mask perfectly amodal.

For the two cameras, I would avoid simply adding the counts from camera A and camera B. Since the cameras observe opposite faces of the same pallet, treat them as two observations of one pallet. Give each pallet a unique track/session ID and fuse the observations using the pallet's travel direction, timestamps and a common counting ROI. The final count should be based on the union of matched instances, not `count_A + count_B`.

Because the cameras are fixed, another option worth testing is line/zone-based counting with tracking rather than relying entirely on segmentation. If each pallet passes through a controlled ROI, this can make the counting logic considerably simpler.

For the edge/TensorRT requirement, I would benchmark the smallest model that meets the accuracy target. Start with `yolov8n-seg`/`yolo11n-seg`, then move to `s` only if the small model cannot handle the occlusion and small-object cases. Also evaluate performance on the actual deployment resolution rather than only on resized training images.

Most importantly, I would first establish a baseline with:
`pretrained seg model → consistent visible masks → BoT-SORT/ByteTrack → ROI counting`

Then add a modal masks and cross-camera fusion one component at a time. That makes it much easier to determine which part actually improves the final pallet count.

**Reply by pderrenger** [reply]
A good baseline is pretrained `yolo11n-seg.pt` or `yolo11s-seg.pt` with consistently labeled **visible** masks and pallet/sequence-level splits; Ultralytics segmentation learns the supplied polygons but has no dedicated amodal-mask mechanism. For non-overlapping opposite views, track within each camera, associate the pallet using timestamps/aisle events, then reconcile pallet-level counts—instance “union” is unreliable without shared appearance or IDs. Benchmark detection versus [instance segmentation](https://docs.ultralytics.com/tasks/segment), add occlusion-heavy samples before model complexity, and validate the final model after [TensorRT export](https://docs.ultralytics.com/integrations/tensorrt) on the target device. Representative frames and annotations from both stacking patterns would help identify the main failure mode.
